From ddfa75cc53ccaf8e5fd029e6516a69ff97e15183 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:29:38 -0400
Subject: [net] openvswitch: Read tcp flags only then the tranport header is present

Message-id: <8d4cf60aea98dd327c630b4f0450880b933b18d3.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94010
O-Subject: [RHEL7.1 net PATCH 05/50] openvswitch: Read tcp flags only then the tranport header is present.
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit 04382a3303c22b0c536fbd0c94c1f012f2b8ed60
Author: Jarno Rajahalme <jrajahalme@nicira.com>
Date:   Sat Feb 15 17:37:45 2014 -0800

    openvswitch: Read tcp flags only then the tranport header is present.

    Only the first IP fragment can have a TCP header, check for this.

    Signed-off-by: Jarno Rajahalme <jrajahalme@nicira.com>
    Signed-off-by: Jesse Gross <jesse@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/flow.c b/net/openvswitch/flow.c
index d79b809..a422471 100644
--- a/net/openvswitch/flow.c
+++ b/net/openvswitch/flow.c
@@ -73,6 +73,7 @@ void ovs_flow_stats_update(struct sw_flow *flow, struct sk_buff *skb)
 
  if ((flow->key.eth.type == htons(ETH_P_IP) ||
       flow->key.eth.type == htons(ETH_P_IPV6)) &&
+     flow->key.ip.frag != OVS_FRAG_TYPE_LATER &&
      flow->key.ip.proto == IPPROTO_TCP &&
      likely(skb->len >= skb_transport_offset(skb) + sizeof(struct tcphdr))) {
   tcp_flags = TCP_FLAGS_BE16(tcp_hdr(skb));
-- 
1.7.1